Allow users to continue onboarding if exited

daquinons opened the pull request:
Fixes #6931 
Requires https://github.com/Automattic/woocommerce-payments-server/pull/3830

#### Changes proposed in this Pull Request
- The PR uses the newly added `submittedDetails` account status parameter to modify the existing `UpdateBusinessDetailsTask` to prompt the user to finish onboarding.
- The new version of the tasks modifies the copy and action, using accounts_link instead of login_link. This allows to continue the onboarding in the step that was left, maintaining any prefilled information sent in the new onboarding flow.
- It introduces the `partially_onboarded` concept, where an account exists, it's connected, but haven't submitted all its details.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

![image](https://github.com/Automattic/woocommerce-payments/assets/2612426/a8c2fa85-7a18-4feb-aad5-74cab68a5403)


#### Testing instructions

This should work in both New onboarding flow (with or without PO) and in the old flow.

[In new onboarding flow]
- After completing the hosted onboarding and being redirected to Stripe, click on return to site without adding your phone number.
- See the new card in the dashboard urging you to finish onboarding
- Click on “Complete registration”
- You are redirected to the “enter mobile phone” screen on Stripe
- Enter the test phone and see the data that you previously added in the hosted onboarding form here (meaning we are effectively recovering the same onboarding that was started)
- Note: clicking on return to site now after having been returned causes locally p1692334055012739-slack-C03KTTK2YMA (still not 100% sure that it’s expected :D so I’d appreciate confirming this in a non-docker environment too)
- After completing the onboarding the card in the dashboard is removed and everything gets back to normal

[In old onboarding flow]
- If leaving the onboarding on the “enter mobile phone” screen, you are redirected to Connect account. This is the existing behaviour and I think it’s the correct one because if there is no prefilled data and no phone number, so doens’t make sense to show the dashboard at this point, better to begin from zero.
- If you enter the phone number and also advance entering country and business structure, clicking on "Return to site" you are redirected to the Payments dashboard, and the card to continue onboarding is shown. At this point, when clicking, you won’t see prefilled data but you will continue in the step where you left it.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-6.4.0#allow-users-to-continue-onboarding-if-exited
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

daquinons committed:
Modify business detail task to continue onboarding

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7009` or branch name `fix/6931-return-site-blocked-account` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/6931-return-site-blocked-account) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 00d9f24d86321f412fd57a35aa36563e853e3c76
- Build time: 2023-08-25 12:12:01 UTC

_Note: the build is updated when a new commit is pushed to this PR._

daquinons committed:
Request account link if account not fully onboarded

dmallory42 commented on includes/class-wc-payments-account.php:
```diff
@@ -584,11 +600,21 @@ public function maybe_redirect_to_server_link() {
 		try {
-			$link = $this->payments_api_client->get_link( $args );
-			if ( isset( $args['type'] ) && 'complete_kyc_link' === $args['type'] && isset( $link['state'] ) ) {
-				set_transient( 'wcpay_stripe_onboarding_state', $link['state'], DAY_IN_SECONDS );
```
I did a quick check and it seems like this transient gets checked in a few places ([ref](https://github.com/Automattic/woocommerce-payments/blob/update/disable-apms-po/includes/class-wc-payments-account.php#L1247)) - would we still need to set it in the case of a `complete_kyc_link` to make sure this logic works properly?

daquinons replied:
Thanks Dan! I didn't intend to remove that set transient, thanks for catching it up :D Readded here https://github.com/Automattic/woocommerce-payments/pull/7009/commits/171bc4aa3f79a1c9cf49170f70332d0991f7d629

daquinons committed:
Add default value to detailsSubmitted

daquinons committed:
Add changelog

daquinons committed:
Readd set transient

daquinons committed:
Pass args to API call

daquinons committed:
Pass args to function

daquinons committed:
Add type validation to link request in API

daquinons committed:
Make default value for detailsSubmitted as true

daquinons committed:
Merge branch 'develop' into fix/6931-return-site-blocked-account

daquinons committed:
Merge branch 'fix/6931-return-site-blocked-account' of https://github.com/Automattic/woocommerce-payments into fix/6931-return-site-blocked-account

daquinons committed:
Add task test

daquinons committed:
Add php tests

daquinons committed:
Pass type properly

daquinons committed:
Merge branch 'develop' into fix/6931-return-site-blocked-account

daquinons committed:
Merge branch 'develop' into fix/6931-return-site-blocked-account

dmallory42 reviewed (approved):
Looks good and works well, thanks for working on this! I think it's a huge improvement that makes life a lot easier for merchants. 😁

I have some notes that we may want to either look at before merging or in follow-up issues, I know we want to fit this in the next release so I'm pre-approving and happy to go with whatever you feel is best. 🙂

**Handling PO accounts**

It works as expected in both new and old onboarding, however I noticed when I tried with a PO account (selecting `US` / `indvidual` / `ready to start selling` / `less than $250k`), when I clicked to go back on the first step of the onboarding, instead of seeing the new task, I saw the PO "start receiving deposits" task:

<img width="736" alt="Screenshot 2023-08-25 at 11 26 42" src="https://github.com/Automattic/woocommerce-payments/assets/11770181/47d4d80a-5075-45e0-8b7e-e2e48d198dfe">

I think it would be better to show your new task _instead_ of this task in the case where a user didn't complete the initial onboarding, since if they haven't completed initial KYC yet they can't even accept payments. 

**Text Updates**

<img width="686" alt="Screenshot 2023-08-25 at 11 34 07" src="https://github.com/Automattic/woocommerce-payments/assets/11770181/4106adee-00d0-4b3a-9a48-8032488684ad">

If possible, I wonder if we can get a design review on the text used? It may be just me, but I don't normally see us using the word 'Registration' when talking to merchants. I was thinking something like `Complete WooPayments Setup` or `Set up WooPayments` as the task name, and then `Payments and deposits are disabled for this account until setup is completed.` for the description. What do you think?

dmallory42 commented:
> Note: clicking on return to site now after having been returned causes locally p1692334055012739-slack-C03KTTK2YMA (still not 100% sure that it’s expected :D so I’d appreciate confirming this in a non-docker environment too)

I'm not sure either if this is expected, but I can confirm that using sandbox and a JT site, the redirect works as expected.

daquinons committed:
Update text task

daquinons committed:
Display task also in PO

daquinons committed:
Merge branch 'fix/6931-return-site-blocked-account' of https://github.com/Automattic/woocommerce-payments into fix/6931-return-site-blocked-account

daquinons commented:
> Looks good and works well, thanks for working on this! I think it's a huge improvement that makes life a lot easier for merchants. 😁
> 
> I have some notes that we may want to either look at before merging or in follow-up issues, I know we want to fit this in the next release so I'm pre-approving and happy to go with whatever you feel is best. 🙂
> 
> **Handling PO accounts**
> 
> It works as expected in both new and old onboarding, however I noticed when I tried with a PO account (selecting `US` / `indvidual` / `ready to start selling` / `less than $250k`), when I clicked to go back on the first step of the onboarding, instead of seeing the new task, I saw the PO "start receiving deposits" task:
> <img alt="Screenshot 2023-08-25 at 11 26 42" width="736" src="https://user-images.githubusercontent.com/11770181/263245046-47d4d80a-5075-45e0-8b7e-e2e48d198dfe.png">
> 
> I think it would be better to show your new task _instead_ of this task in the case where a user didn't complete the initial onboarding, since if they haven't completed initial KYC yet they can't even accept payments.
> 

Thank you @dmallory42 ! I've addressed it here https://github.com/Automattic/woocommerce-payments/pull/7009/commits/3e8fc77b890022c77682f9af934f9225204a4a4a
> **Text Updates** 
> <img alt="Screenshot 2023-08-25 at 11 34 07" width="686" src="https://user-images.githubusercontent.com/11770181/263245828-4106adee-00d0-4b3a-9a48-8032488684ad.png">
> 
> If possible, I wonder if we can get a design review on the text used? It may be just me, but I don't normally see us using the word 'Registration' when talking to merchants. I was thinking something like `Complete WooPayments Setup` or `Set up WooPayments` as the task name, and then `Payments and deposits are disabled for this account until setup is completed.` for the description. What do you think?

Thank you! I agree, I've updated it https://github.com/Automattic/woocommerce-payments/pull/7009/commits/88df7147b098bc45a818e2284c854548222731b2

<img width="702" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/2612426/92502880-28ab-49ff-9c28-dc4d23aa2710">

daquinons committed:
Change task display

daquinons commented:
@dmallory42 Actually this one https://github.com/Automattic/woocommerce-payments/pull/7009/commits/16d955e1814775272888da15389da87a44e2caa1 the previous change was going to display the business update details in PO after details are submitted.

daquinons committed:
Display PO task only after details submitted

daquinons committed:
Fix tests

daquinons committed:
Fix test
